modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java

```diff
@@ -105,6 +106,13 @@ public interface EntryReader {
     /** Failure processor that is used to handle critical errors. */
     private final FailureManager failureManager;
 
+    /**
+     * Whether a failure in notification chain was passed to the FailureHandler. Used to make sure that we only pass first such a failure
+     * because, as any failure in the chain will stop any notifications, it only makes sense to log the first one. Subsequent ones will
+     * be instances of the same original exception.
+     */
+    private final AtomicBoolean firedFailureOnChain = new AtomicBoolean(false);
```

Comment:
`false` is not necessary, it is up to you.

Reply:
In this case it seems good to have it explicit